hera_cal/vis_clean.py

```diff
@@ -0,0 +1,471 @@
+                                                      filt2d_mode=filt2d_mode)
+                    flgs = np.zeros_like(mdl, dtype=np.bool)
+                else:
+                    # flagged baseline
+                    mdl = np.zeros_like(d)
+                    res = d - mdl
+                    flgs = np.ones_like(mdl, dtype=np.bool)
+                    info = {'skipped': True}
+
+            # append to new Containers
+            self.clean_model[k] = mdl
+            self.clean_resid[k] = res
+            self.clean_flags[k] = flgs
+            self.clean_info[k] = info
+
+    def fft_data(self, data=None, flags=None, keys=None, assign='dfft', ax='freq', window='none', alpha=0.1,
+                 overwrite=False, edgecut_low=0, edgecut_hi=0, ifft=True, fftshift=True, verbose=True):
+        """
+        Take FFT of data and attach to self via assign kwarg. Default is self.dfft.
+        Take note of the adopted fourier convention via ifft and fftshift kwargs.
+
+        Args:
+            data : DataContainer
+                Object to pull data to FT from. Default is self.data.
+            flags : DataContainer
+                Object to pull flags in FT from. Default is no flags.
+            keys : list of tuples
+                List of keys from clean_data to FFT. Default is all keys.
+            assign : str
+                Name of DataContainer to attach to self. Default is self.dfft
+            ax : str, options=['freq', 'time', 'both']
+                Axis along with to take FFT.
+            window : str
+                Windowing function to apply across frequency before FFT. If ax is 'both',
+                can feed as a tuple specifying window for 0th and 1st FFT axis.
+            alpha : float
+                If window is 'tukey' this is its alpha parameter. If ax is 'both',
+                can feed as a tuple specifying alpha for 0th and 1st FFT axis.
+            edgecut_low : int, number of bins to consider zero-padded at low-side of the FFT axis,
+                such that the windowing function smoothly approaches zero. If ax is 'both',
+                can feed as a tuple specifying for 0th and 1st FFT axis.
+            edgecut_hi : int, number of bins to consider zero-padded at high-side of the FFT axis,
+                such that the windowing function smoothly approaches zero. If ax is 'both',
+                can feed as a tuple specifying for 0th and 1st FFT axis.
+            ifft : bool, if True, use ifft instead of fft
+            fftshift : bool, if True, fftshift along FFT axes.
+            overwrite : bool
+                If dfft[key] already exists, overwrite its contents.
+        """
+        # type checks
+        assert ax in ['freq', 'time', 'both'], "ax must be one of ['freq', 'time', 'both']"
+        fft2d = ax == 'both'
+        if fft2d:
+            # 2D fft
+            if isinstance(window, (str, np.str)):
```

Comment:
There is no explicit typechecking of `window` before this statement, and the code does nothing if it's not. For instance, if the user passes in `None`, the code on line 411 will try to call:
```
w1 = dspec.gen_window(None[0], ...)
```
which will give a confusing `TypeError: 'NoneType' object has no attribute '__getitem__'`. I think you should either make these arguments into 2-tuples unconditionally and punt on error handling to the `gen_window` function, or enforce the type checking outside of this conditional (e.g., you should be checking that `window` is a string even for the 1d CLEAN case). I understand that the former is less work, but I'd advocate doing the latter so you can give the user more direct error messages.

Reply:
That's actually not right. In the case of `fft2d`, window actually becomes `(window, window)` (line 392), so what's passed to `dspec.gen_window` is `(None, None)[0]`, and within `dspec.gen_window` there is type checking for None values (among others).

Reply:
The code, as written here, will only make it `(window, window)` if the `window` kwarg is string-like (`str` or `np.str`). There is no `else` clause, and so if the user passes in `window=None`, it will stay `None`, and not be turned into `(None, None)`. What I'm saying is that if you _do_ want to punt to the `gen_window` error handling, this tuple-fication should be done unconditionally. That is, you should not be using the `isinstance` check, and just assign `window = (window, window)`. Otherwise you'll get strange indexing errors if the arguments are not the expected types.

Reply:
ah you're right, sorry. I should have written: `if not isinstance(window, collections.Iterable)`